File: Backend/src/main/java/com/sencarmarket/module/certification/mapper/CertificationMapper.java
package com.sencarmarket.module.certification.mapper;

import com.sencarmarket.module.certification.dto.DemandeCertificationResponse;
import com.sencarmarket.module.certification.dto.InspectionResponse;
import com.sencarmarket.module.certification.dto.RapportInspectionResponse;
import com.sencarmarket.module.certification.entity.DemandeCertification;
import com.sencarmarket.module.certification.entity.Inspection;
import com.sencarmarket.module.certification.entity.RapportInspection;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Mapper pour la conversion entre entités et DTOs du module Certification
 * Suit le principe DRY en centralisant les conversions
 */
@Component
public class CertificationMapper {

    /**
     * Convertit une entité DemandeCertification en DemandeCertificationResponse
     */
    public DemandeCertificationResponse toDemandeCertificationResponse(DemandeCertification demande) {
        if (demande == null) {
            return null;
        }
        return DemandeCertificationResponse.builder()
                .id(demande.getId())
                .utilisateurId(demande.getUtilisateur() != null ? demande.getUtilisateur().getId() : null)
                .utilisateurNom(demande.getUtilisateur() != null ? demande.getUtilisateur().getNom() : null)
                .vehiculeId(demande.getVehicule() != null ? demande.getVehicule().getId() : null)
                .vehiculeDescription(demande.getVehicule() != null ? 
                        demande.getVehicule().getMarque() + " " + demande.getVehicule().getModele() : null)
                .statut(demande.getStatut())
                .montantPaiement(demande.getMontantPaiement())
                .paiementId(demande.getPaiementId())
                .inspecteurId(demande.getInspecteur() != null ? demande.getInspecteur().getId() : null)
                .inspecteurNom(demande.getInspecteur() != null ? demande.getInspecteur().getNom() : null)
                .dateSoumission(demande.getDateSoumission())
                .dateTraitement(demande.getDateTraitement())
                .dateInspection(demande.getDateInspection())
                .motifRejet(demande.getMotifRejet())
                .badgeCertifieUrl(demande.getBadgeCertifieUrl())
                .createdAt(demande.getCreatedAt())
                .updatedAt(demande.getUpdatedAt())
                .build();
    }

    /**
     * Convertit une liste d'entités DemandeCertification en liste de DemandeCertificationResponse
     */
    public List<DemandeCertificationResponse> toDemandeCertificationResponseList(List<DemandeCertification> demandes) {
        if (demandes == null) {
            return null;
        }
        return demandes.stream()
                .map(this::toDemandeCertificationResponse)
                .collect(Collectors.toList());
    }

    /**
     * Convertit une entité Inspection en InspectionResponse
     */
    public InspectionResponse toInspectionResponse(Inspection inspection) {
        if (inspection == null) {
            return null;
        }
        return InspectionResponse.builder()
                .id(inspection.getId())
                .demandeCertificationId(inspection.getDemandeCertification() != null ? 
                        inspection.getDemandeCertification().getId() : null)
                .inspecteurId(inspection.getInspecteur() != null ? inspection.getInspecteur().getId() : null)
                .inspecteurNom(inspection.getInspecteur() != null ? inspection.getInspecteur().getNom() : null)
                .dateInspection(inspection.getDateInspection())
                .resultat(inspection.getResultat())
                .commentaire(inspection.getCommentaire())
                .kilometrage(inspection.getKilometrage())
                .etatMoteur(inspection.getEtatMoteur())
                .etatGenerateur(inspection.getEtatGenerateur())
                .etatFreinage(inspection.getEtatFreinage())
                .etatSuspension(inspection.getEtatSuspension())
                .etatTransmission(inspection.getEtatTransmission())
                .etatPneus(inspection.getEtatPneus())
                .etatCarrosserie(inspection.getEtatCarrosserie())
                .etatInterieur(inspection.getEtatInterieur())
                .scoreTotal(inspection.getScoreTotal())
                .createdAt(inspection.getCreatedAt())
                .updatedAt(inspection.getUpdatedAt())
                .build();
    }

    /**
     * Convertit une liste d'entités Inspection en liste de InspectionResponse
     */
    public List<InspectionResponse> toInspectionResponseList(List<Inspection> inspections) {
        if (inspections == null) {
            return null;
        }
        return inspections.stream()
                .map(this::toInspectionResponse)
                .collect(Collectors.toList());
    }

    /**
     * Convertit une entité RapportInspection en RapportInspectionResponse
     */
    public RapportInspectionResponse toRapportInspectionResponse(RapportInspection rapport) {
        if (rapport == null) {
            return null;
        }
        return RapportInspectionResponse.builder()
                .id(rapport.getId())
                .inspectionId(rapport.getInspection() != null ? rapport.getInspection().getId() : null)
                .urlRapportPdf(rapport.getUrlRapportPdf())
                .dateGeneration(rapport.getDateGeneration())
                .scoreGlobale(rapport.getScoreGlobale())
                .recommendations(rapport.getRecommendations())
                .conclusion(rapport.getConclusion())
                .estApprouve(rapport.getEstApprouve())
                .createdAt(rapport.getCreatedAt())
                .updatedAt(rapport.getUpdatedAt())
                .build();
    }

    /**
     * Convertit une liste d'entités RapportInspection en liste de RapportInspectionResponse
     */
    public List<RapportInspectionResponse> toRapportInspectionResponseList(List<RapportInspection> rapports) {
        if (rapports == null) {
            return null;
        }
        return rapports.stream()
                .map(this::toRapportInspectionResponse)
                .collect(Collectors.toList());
    }
}

File: Backend/src/modules/certification/services/certification.mapper.ts
import { Injectable } from "@nestjs/common";

import { toNullableNumber } from "../../../common/utils/number.util";
import {
  DemandeRecord,
  InspectionRecord,
  RapportRecord,
  VehiculeMini,
} from "../certification.models";
import { DemandeCertificationResponseDto } from "../dto/demande-certification-response.dto";
import { InspectionResponseDto } from "../dto/inspection-response.dto";
import { RapportInspectionResponseDto } from "../dto/rapport-inspection-response.dto";

@Injectable()
export class CertificationMapper {
  toDemandeResponse(demande: DemandeRecord): DemandeCertificationResponseDto {
    return {
      id: demande.id,
      utilisateurId: demande.utilisateur_id,
      utilisateurNom:
        demande.utilisateur_demande_certification_utilisateur_idToutilisateur
          ?.nom ?? null,
      vehiculeId: demande.vehicule_id,
      vehiculeDescription: this.vehiculeDescription(demande.vehicule),
      statut: demande.statut,
      montantPaiement: toNullableNumber(demande.montant_paiement),
      paiementId: demande.paiement_id,
      inspecteurId: demande.inspecteur_id,
      inspecteurNom:
        demande.utilisateur_demande_certification_inspecteur_idToutilisateur
          ?.nom ?? null,
      dateSoumission: demande.date_soumission,
      dateTraitement: demande.date_traitement,
      dateInspection: demande.date_inspection,
      motifRejet: demande.motif_rejet,
      badgeCertifieUrl: demande.badge_certifie_url,
      createdAt: demande.created_at,
      updatedAt: demande.updated_at,
    };
  }

  toInspectionResponse(inspection: InspectionRecord): InspectionResponseDto {
    return {
      id: inspection.id,
      demandeCertificationId: inspection.demande_certification_id,
      inspecteurId: inspection.utilisateur?.id ?? inspection.inspecteur_id,
      inspecteurNom: inspection.utilisateur?.nom ?? null,
      dateInspection: inspection.date_inspection,
      resultat: inspection.resultat,
      commentaire: inspection.commentaire,
      kilometrage: inspection.kilometrage,
      etatMoteur: inspection.etat_moteur,
      etatGenerateur: inspection.etat_generateur,
      etatFreinage: inspection.etat_freinage,
      etatSuspension: inspection.etat_suspension,
      etatTransmission: inspection.etat_transmission,
      etatPneus: inspection.etat_pneus,
      etatCarrosserie: inspection.etat_carrosserie,
      etatInterieur: inspection.etat_interieur,
      scoreTotal: inspection.score_total,
      createdAt: inspection.created_at,
      updatedAt: inspection.updated_at,
    };
  }

  toRapportResponse(rapport: RapportRecord): RapportInspectionResponseDto {
    return {
      id: rapport.id,
      inspectionId: rapport.inspection_id,
      urlRapportPdf: rapport.url_rapport_pdf,
      dateGeneration: rapport.date_generation,
      scoreGlobale: rapport.score_globale,
      recommendations: rapport.recommendations,
      conclusion: rapport.conclusion,
      estApprouve: rapport.est_approuve,
      createdAt: rapport.created_at,
      updatedAt: rapport.updated_at,
    };
  }

  private vehiculeDescription(vehicule: VehiculeMini): string | null {
    const desc =
      `${vehicule.marque?.nom ?? ""} ${vehicule.modele?.nom ?? ""}`.trim();
    return desc || null;
  }
}
